### PR TITLE
Add Tuple2K Semigroupal

### LINKED ISSUE #3410: Missing TupleK Semigroupal
I don't see an implementation (although there is ContravariantMonoidal, which is more constrained).

I'll plan to send a PR, but if anyone sees a reason why this can't exist, please feel free to let me know.

### PR DESCRIPTION
**Pull request by rmehri01**
Closes #3410.

I wasn't really sure where in the Tuple2kInstances this should be added so I just put it in the last one so please tell me if there is somewhere else it should go. Also, should tests be added because I had a bit of trouble doing so. Thanks!

### CODE REVIEW

File: core/src/main/scala/cats/data/Tuple2K.scala [hunk 1]
```diff
@@ -189,6 +189,22 @@ sealed abstract private[data] class Tuple2KInstances8 {
       def F: Functor[F] = FF
       def G: Functor[G] = GG
     }
+  implicit def catsDataSemigroupalForTuple2K[F[_], G[_]](implicit
+    FF: Semigroupal[F],
+    GG: Semigroupal[G]
+  ): Semigroupal[λ[α => Tuple2K[F, G, α]]] =
```
Comment (LukaJCB) [inline core/src/main/scala/cats/data/Tuple2K.scala:195]: any reason for not using the shorthand `*` here? :)

Comment (LukaJCB) [reply core/src/main/scala/cats/data/Tuple2K.scala:195]: Oh I see, it's not used in any of the others either, that's odd, but should be fine :)

Comment (rmehri01) [reply core/src/main/scala/cats/data/Tuple2K.scala:195]: Yeah I was just trying to follow the style of the others, unless you want me to change them?

Comment (LukaJCB) [reply core/src/main/scala/cats/data/Tuple2K.scala:195]: no, it's all good! thanks!

**Comment by rmehri01** [comment]
Alright I can do that. Just another thing, does it matter which tuple2k instance class this goes in because there was 0-8 so I wasn't really sure and should this be tested?

File: core/src/main/scala/cats/data/Tuple2K.scala (hunk 1 above)
Comment (rmehri01) [reply core/src/main/scala/cats/data/Tuple2K.scala:195]: Alright sounds good :)

**Comment by rmehri01** [comment]
Also is that import at the top of the file `import cats.Contravariant` needed? Intellij says it isn't used.

**Comment by joroKr21** [comment]
> Alright I can do that. Just another thing, does it matter which tuple2k instance abstract class this goes in because there was 0-8 so I wasn't really sure and also should this be tested?

This is the correct place. I'm not sure about testing, I guess you would need a type that has `Semigroupal` but not `Applicative` to test this.

**Comment by rmehri01** [comment]
> > Alright I can do that. Just another thing, does it matter which tuple2k instance abstract class this goes in because there was 0-8 so I wasn't really sure and also should this be tested?
> 
> This is the correct place. I'm not sure about testing, I guess you would need a type that has `Semigroupal` but not `Applicative` to test this.

Ah okay thanks for the response, I meant more like does it matter which tuple2k instances class this goes in because I'm not really sure why they are labelled 0-8.

**Comment by joroKr21** [comment]
> Also does it matter which tuple2k instances class this goes in because I'm not really sure why they are labelled 0-8, I thought maybe it was because they had some sort of structure.

It matters but this is the correct place to put it. The reason they exist is implicit priority. `Semigroupal` should have the lowest priority (same as `Functor`) because it has the least constraints. There is also no overlap between `Functor` and `Semigroupal` so they can be in one place without clashing.

**Comment by rmehri01** [comment]
> It matters but this is the correct place to put it. The reason they exist is implicit priority. `Semigroupal` should have the lowest priority (same as `Functor`) because it has the least constraints. There is also no overlap between `Functor` and `Semigroupal` so they can be in one place without clashing.

Ah gotcha, thank you! :)

**Comment by rmehri01** [comment]
Np, thanks for the help everyone! :)

**Pull request merged**